oscore-cf/src/main/java/org/eclipse/californium/oscore/HashMapCtxDB.java

```diff
@@ -0,0 +1,243 @@
+/*******************************************************************************
+ * Copyright (c) 2018 RISE SICS and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Joakim Brorsson
+ *    Ludwig Seitz (RISE SICS)
+ *    Tobias Andersson (RISE SICS)
+ *    
+ ******************************************************************************/
+package org.eclipse.californium.oscore;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.eclipse.californium.core.coap.Token;
+
+/**
+ * 
+ * Implements the OSCoreCtxDB interface with HashMaps.
+ *
+ */
+public class HashMapCtxDB implements OSCoreCtxDB {
+
+	/**
+	 * The logger
+	 */
+	private static final Logger LOGGER = LoggerFactory.getLogger(HashMapCtxDB.class.getName());
+
+	private HashMap<ByteId, OSCoreCtx> ridMap;
+	private HashMap<Token, OSCoreCtx> tokenMap;
+	private HashMap<String, OSCoreCtx> uriMap;
+	private HashMap<Token, Integer> seqMap;
+
+	private ArrayList<Token> allTokens;
+
+	private static volatile HashMapCtxDB singleton = null;
+
+	/**
+	 * Create the database
+	 */
+	private HashMapCtxDB() {
+
+		// Prevent form the reflection api.
+		if (singleton != null) {
+			throw new RuntimeException("Use getInstance() method to get the single instance of this class.");
+		}
+
+		this.tokenMap = new HashMap<>();
+		this.ridMap = new HashMap<>();
+		this.uriMap = new HashMap<>();
+		this.seqMap = new HashMap<>();
+		this.allTokens = new ArrayList<Token>();
+	}
+
+	/**
+	 * @return the singleton instance of this context database
+	 */
+	public static HashMapCtxDB getInstance() {
```

Comment:
Should all `CoapEndpoint`s share the same CtxDB? I'm not sure, if the "unique" grants (e.g. Token) will also match for different receiving/sending `CoapEndpoint`s.

Reply:
We have not talked much about the multiple endpoint situation. The reason why the db was made into a singleton was to avoid passing it along with the exchange object. Your point is good and I think this needs some discussion.